mocpy/moc.py

```diff
@@ -434,147 +223,112 @@ def from_coo_list_no_iteration(cls, skycoords, max_norder):
-            raise ValueError('format should be one of %s' % (str(formats)))
-        
-        self._ensure_consistency()
-        uniq_array = []
-        for uniq in self.uniq_pixels_iterator():
-            uniq_array.append(uniq)
-        
-        if format == 'fits':
-            if self._order is not None:
-                moc_order = self._order
-            else:
-                moc_order = self.max_order
-            if moc_order<=13:
-                format = '1J'
-            else:
-                format = '1K'
-            
-            tbhdu = fits.BinTableHDU.from_columns(fits.ColDefs([fits.Column(name='UNIQ', format=format, array=np.array(uniq_array))]))
-            tbhdu.header['PIXTYPE'] = 'HEALPIX'
-            tbhdu.header['ORDERING'] = 'NUNIQ'
-            tbhdu.header['COORDSYS'] = 'C'
-            tbhdu.header['MOCORDER'] = moc_order
-            tbhdu.header['MOCTOOL'] = 'MOCPy'
-            if optional_kw_dict:
-                for key in optional_kw_dict:
-                    tbhdu.header[key] = optional_kw_dict[key]
-                    
-            thdulist = fits.HDUList([fits.PrimaryHDU(), tbhdu])
-            thdulist.writeto(path, clobber=True)
-            
-        elif format == 'json':
-            import json
-            
-            json_moc = {}
-            pix_list = []
-            o_order = -1
-            for uniq in uniq_array:
-                order, ipix = utils.uniq2orderipix(uniq)
-                if order!=o_order and o_order>0:
-                    json_moc[str(o_order)] = pix_list
-                    pix_list = []
-                
-                o_order = order
-                pix_list.append(ipix)
-                
-            json_moc[str(o_order)] = pix_list
-            with open(path, 'w') as h:
-                h.write(json.dumps(json_moc, sort_keys=True, indent=2))
+        return self._query(table_id, max_rows)
 
+    def _query(self, resource_id, max_rows):
+        """
+        internal method to query Simbad or a VizieR table
+        for sources in the coverage of the MOC instance
+        """
+
+        if max_rows is not None and max_rows >= 0:
+            max_rows_str = str(max_rows)
+        else:
+            max_rows_str = str(9999999999)
+
+        tmp_moc = tempfile.NamedTemporaryFile(delete=False)
+        self.write(tmp_moc.name)
+        r = requests.post('http://cdsxmatch.u-strasbg.fr/QueryCat/QueryCat',
+                          data={'mode': 'mocfile',
+                                'catName': resource_id,
+                                'format': 'votable',
+                                'limit': max_rows_str},
+                          files={'moc': open(tmp_moc.name, 'rb')},
+                          headers={'User-Agent': 'MOCPy'},
+                          stream=True)
         
-class MOC_io:
-    @staticmethod
-    def read_local(path, moc_order=None):
+        tmp_vot = tempfile.NamedTemporaryFile(delete=False)
```

Comment:
Maybe create a [BytesIO](https://pythonhosted.org/six/#six.BytesIO) and write / read there? Should be more efficient than creating a temp file on disc and reading / writing there, no?
Note that `BytesIO` objects behave like files and should just work with `parse_single_table`.
Example: https://github.com/astropy/astropy/blob/74a8c7a22dbcbcb18945d521dd0c77de787ce43a/astropy/io/votable/tests/table_test.py#L151

Reply:
Thanks for that. Seems to be really faster now !